src/ccl_power.c

```diff
@@ -436,124 +436,134 @@ static void ccl_cosmology_compute_power_class(ccl_cosmology * cosmo, int * statu
   }
-  else{
-    // After this loop x will contain log(k), y will contain log(P_nl), z will contain log(P_lin)
+
+  //If not, proceed
+  if(!*status){
+    
+    // After this loop x will contain log(k)
     // all in Mpc, not Mpc/h units!
     double psout_l,ic;
-    int s=0;
+    int newstatus=0;
     for (int i=0; i<nk; i++) {
       for (int j = 0; j < na; j++) {
 	//The 2D interpolation routines access the function values y_{k_ia_j} with the following ordering:
 	//y_ij = y2d[j*N_k + i]
 	//with i = 0,...,N_k-1 and j = 0,...,N_a-1.
-	s |= spectra_pk_at_k_and_z(&ba, &pm, &sp,x[i],1./z[j]-1., &psout_l,&ic);
+	newstatus |= spectra_pk_at_k_and_z(&ba, &pm, &sp,x[i],1./a[j]-1., &psout_l,&ic);
 	y2d_lin[j*nk+i] = log(psout_l);
       }
       x[i] = log(x[i]);
     }
-    if(s) {
-      free(x);
-      free(z);
-      free(y2d_nl);
-      free(y2d_lin);
+
+    //If error, store status, we will free later
+    if(newstatus) {
       *status = CCL_ERROR_CLASS;
       ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): Error computing CLASS power spectrum\n");
+    }
 
-      ccl_free_class_structs(cosmo, &ba,&th,&pt,&tr,&pm,&sp,&nl,&le,init_arr,status);
+    //If not proceed
+    if(!newstatus){
+      
+      gsl_spline2d * log_power = gsl_spline2d_alloc(PLIN_SPLINE_TYPE, nk,na);
+      pwstatus = gsl_spline2d_init(log_power, x, a, y2d_lin,nk,na);
 
-      return;
-    }
-    gsl_spline2d * log_power = gsl_spline2d_alloc(PLIN_SPLINE_TYPE, nk,na);
-    int pwstatus = gsl_spline2d_init(log_power, x, z, y2d_lin,nk,na);
-    if (pwstatus) {
-      free(x);
-      free(z);
-      free(y2d_nl);
-      free(y2d_lin);
-      gsl_spline2d_free(log_power);
-      ccl_free_class_structs(cosmo, &ba,&th,&pt,&tr,&pm,&sp,&nl,&le,init_arr,status);
-      *status = CCL_ERROR_SPLINE;
-      ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): Error creating log_power spline\n");
-      return;
-    }
-    else {
-      cosmo->data.p_lin = log_power;
-    }
+      //If error, store status, we will free later
+      if (pwstatus) {
+	*status = CCL_ERROR_SPLINE;
+	ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): Error creating log_power spline\n");
+      }
 
-    // At the moment KMIN can't be less than CLASS's kmin in the nonlinear case.
-    if (kmin<(exp(sp.ln_k[0]))) {
-      *status = CCL_ERROR_CLASS;
-      ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): K_MIN is less than CLASS's kmin. Not yet supported for nonlinear P(k).\n");
+      //If not, proceed
+      if(!pwstatus){
+	cosmo->data.p_lin = log_power;
+      }
     }
+    
+  } else { //Linear power spec failed, so we return without proceeding to nonlinear.
+
+    free(x);
+    free(a);
+    free(y2d_nl);
+    free(y2d_lin);
+    ccl_free_class_structs(cosmo, &ba,&th,&pt,&tr,&pm,&sp,&nl,&le,init_arr,status);
+    return;
+  }
```

Comment:
I think you probably see the pattern now, so I'll refrain from annotating the rest of the function, but the basic pattern is:

start with *status = 0

if (!*status) {  // no errors yet.
    Do something that might cause an error;
    If (there was an error) *status = ERROR_CODE;
}

.. repeat as many times as needed

// At the end clean up any memory
free stuff that was temporary within this function
if (status) {
   free stuff that we aren't returning because of the error
}

return

Reply:
Ok, thanks. Please check the new structure of that function. It's proving more difficult for the other ones and I'm not sure it's necessary to get all these changes in right now. Could we open an issue for the other power spectrum methods?